### PR TITLE
Bearer Token Resolver Configuration

### PR DESCRIPTION
**Pull request by jzheaux**
This introduces #bearerTokenResolver(BearerTokenResolver) to the
Resource Server DSL, allowing users to configure the resolver to allow
the access token as part of the request body or a query parameter. It
also allows the user to replace the resolver with a completely custom
one.

This also introduces the same ability by exposing a bean of type
BearerTokenResolver

Fixes: gh-5479

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/OAuth2Configurer.java [hunk 1]
```diff
@@ -96,7 +97,8 @@ private void initClientConfigurer() {
 	}
 
 	private void initResourceServerConfigurer() {
-		this.resourceServerConfigurer = new OAuth2ResourceServerConfigurer<>();
+		ApplicationContext context = getBuilder().getSharedObject(ApplicationContext.class);
+		this.resourceServerConfigurer = new OAuth2ResourceServerConfigurer<>(context);
```
Comment (jzheaux) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/OAuth2Configurer.java]: @jgrandja you and I were talking about trying to keep the ApplicationContext resolution approach consistent across OAuth2 configurers; however with this set of tests for bearer token resolver, it still seemed the cleanest approach. Let me know if you see one you'd prefer more.

Comment (jgrandja) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/OAuth2Configurer.java]: For `oauth2-client` we use `OAuth2ClientConfigurerUtils` to get `@Bean`'s from the context - passing in the builder (http). This centralizes the logic for `@Bean` lookup so it can be reused across configurer's since there is more than one for `oauth2-client`. I'm not sure if there will be more than one for resource server. Either way, I think it's fine leaving as is.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 2]
```diff
@@ -564,7 +667,7 @@ public void requestWhenCustomJwtDecoderExposedAsBeanThenUsed()
 	@Test
 	public void getJwtDecoderWhenConfiguredWithDecoderAndJwkSetUriThenLastOneWins() {
 		OAuth2ResourceServerConfigurer.JwtConfigurer jwtConfigurer =
-				new OAuth2ResourceServerConfigurer().new JwtConfigurer(null);
+				new OAuth2ResourceServerConfigurer(null).jwt();
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: We should not pass `null` context.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 3]
```diff
@@ -574,7 +677,7 @@ public void getJwtDecoderWhenConfiguredWithDecoderAndJwkSetUriThenLastOneWins()
 		assertThat(jwtConfigurer.getJwtDecoder()).isEqualTo(decoder);
 
 		jwtConfigurer =
-				new OAuth2ResourceServerConfigurer().new JwtConfigurer(null);
+				new OAuth2ResourceServerConfigurer(null).jwt();
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: We should not pass `null` context.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 4]
```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
...
+		GenericWebApplicationContext context = new GenericWebApplicationContext();
+		context.registerBean("resolverOne", BearerTokenResolver.class, () -> resolverBean);
+		context.registerBean("resolverTwo", BearerTokenResolver.class, () -> resolverBean);
+		this.spring.context(context).autowire();
+
+		OAuth2ResourceServerConfigurer oauth2 = new OAuth2ResourceServerConfigurer(context);
+
+		oauth2.bearerTokenResolver(resolver);
+
+		assertThat(oauth2.getBearerTokenResolver()).isEqualTo(resolver);
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: I see why you made `getBearerTokenResolver()` package private. However, I'd rather avoid opening up methods so it can work for testing.

This can be tested by issuing a request and then `verify()` or `verifyZeroInteractions()` to assert which resolver is in effect.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 5]
```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
 		assertThat(result.getRequest().getSession(false)).isNotNull();
 	}
 
+	// -- custom bearer token resolver
+
+	@Test
+	public void requestWhenBearerTokenResolverAllowsRequestBodyThenEitherHeaderOrRequestBodyIsAccepted()
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Please add a test where the token is in the header and body AND another test where the token is in header and query param

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 6]
```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
...
+		BearerTokenResolver resolverBean = mock(BearerTokenResolver.class);
+
+		GenericWebApplicationContext context = new GenericWebApplicationContext();
+		context.registerBean("resolverOne", BearerTokenResolver.class, () -> resolverBean);
+		context.registerBean("resolverTwo", BearerTokenResolver.class, () -> resolverBean);
+		this.spring.context(context).autowire();
+
+		OAuth2ResourceServerConfigurer oauth2 = new OAuth2ResourceServerConfigurer(context);
+
+		assertThatCode(() -> oauth2.getBearerTokenResolver())
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Related to previous comment on making `getBearerTokenResolver()` private...this can be tested by initializing the context which will initialize the configurer and the exception and you can `assertThatThrownBy()`

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 7]
```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
...
+		BearerTokenResolver resolverBean = mock(BearerTokenResolver.class);
+
+		GenericWebApplicationContext context = new GenericWebApplicationContext();
+		context.registerBean(BearerTokenResolver.class, () -> resolverBean);
+		this.spring.context(context).autowire();
+
+		OAuth2ResourceServerConfigurer oauth2 = new OAuth2ResourceServerConfigurer(context);
+		oauth2.bearerTokenResolver(resolver);
+
+		assertThat(oauth2.getBearerTokenResolver()).isEqualTo(resolver);
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: This can be tested by issuing a request and then `verify()` or `verifyZeroInteractions()` to assert which resolver is in effect.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 8]
```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
...
+	}
+
+	@Test
+	public void getBearerTokenResolverWhenNoResolverSpecifiedThenTheDefaultIsUsed() {
+		ApplicationContext context =
+				this.spring.context(new GenericWebApplicationContext()).getContext();
+
+		OAuth2ResourceServerConfigurer oauth2 = new OAuth2ResourceServerConfigurer(context);
+
+		assertThat(oauth2.getBearerTokenResolver()).isInstanceOf(DefaultBearerTokenResolver.class);
```
Comment (jgrandja) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: This can be tested by issuing a request and then ensuring the request works and therefore the default resolver is set or else an error would happen.

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java [hunk 9]
```diff
@@ -231,13 +242,20 @@ private void registerDefaultCsrfOverride(H http) {
 		csrf.ignoringRequestMatchers(this.requestMatcher);
 	}
 
-	private BearerTokenResolver getBearerTokenResolver() {
+	BearerTokenResolver getBearerTokenResolver() {
```
[the remark this answers is not in the document]
Comment (jzheaux) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java]: This needs to be default so that unit tests can access it. What this comes down to is how we prefer to do unit tests for configurers. In the PR for JwtDecoder, the same tradeoff was made in order to allow for corner cases to be confirmed (as unit tests) instead of as performing a full MockMvc roundtrip.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 4 above)
Comment (jzheaux) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Can you elaborate on how this confirms the intended behavior? That is, we want to confirm that the correct implementation is picked. We won't actually invoke the mock without performing some kind of mock mvc roundtrip, which, this being a unit test, might be a bit odd.

We did the same thing with `getJwtDecoder` in `JwtConfigurer` to achieve what appears to me to be following the same reasoning, e.g. corner case should be a unit test, so we make the method package-private so the unit test can confirm. I agree that bumping up the access level of a method for a unit test is worth thinking twice about, but I don't see another way to do it other than reflection or turning it into an integration test.

Now, we could look at this becoming an integration test; however, that seems to deviate from the original thrust of this set of tests. (Again, for context, take a look at the tests for JwtConfigurer and see that some are integration tests and some are unit tests.)

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 5 above)
Comment (jzheaux) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: We have unit tests for these. For my education, how do you gauge what use cases need both an integration test as well as a unit test?

Where my mind goes right now is these integration tests confirm that things got wired correctly and behave accordingly, and unit tests confirm that, given XYZ arbitrary wirings, the contract is correctly upheld.

I don't have a problem adding more tests, I just want to learn from you how you prioritize your own efforts with things like this.

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java]: What's the disadvantage of doing MockMvc roundtrip? I see `SecurityConfigurers` more like integration testing instead of unit testing.

> What this comes down to is how we prefer to do unit tests for configurers

Can you clarify what you mean by "we prefer"? Did I miss a convo you and @rwinch had?

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 5 above)
Comment (jzheaux) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Also, this specific recommendation actually uncovered a bug (awesome!) which the unit test would not have uncovered, which is why I'm interested in your thinking around it.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 2 above)
Comment (jzheaux) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Thank you for catching that! Intended clean up that I missed.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 4 above)
Comment (jgrandja) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: This class mixes unit tests and integration tests which we typically separate in different classes. Also, `SecurityConfigurer`s are integration testing. Are there other configuers that do unit testing with the exception of `JwtConfigurer`?

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java (hunk 5 above)
Comment (jgrandja) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: Sorry I missed that - the unit tests cover the scenarios I suggested. Those tests should be in the units tests.

> Where my mind goes right now is these integration tests confirm that things got wired correctly and behave accordingly

Agreed, the integration tests should confirm things got wired correctly and behave as expected.
